File: 2023/07/ocaml/07.ml
#require "core.top"
#require "ppx_deriving.show"
#require "ppx_deriving.ord"
#require "ppx_sexp_conv"

open Core
open Core.String
open Core.Int
open Core.Map

module Card = struct
  type card =
    | A
    | K
    | Q
    | J
    | T
    | Num of int
  [@@deriving show, sexp_of]     

  let int_of_card card = 
    match card with
    | A -> 14
    | K -> 13
    | Q -> 12
    | J -> 11
    | T -> 10
    | Num x -> x

  let compare c1 c2 = Int.compare (int_of_card c1) (int_of_card c2)

  let rec tiebreaker (c1: card list) (c2: card list) : int = 
    match c1, c2 with
    | hd1 :: tl1, hd2 :: tl2 -> (
      let check = compare hd1 hd2 in
      if check = 0 then (tiebreaker tl1 tl2) else check
    )
    | _ -> 0

  type t = card [@@deriving sexp_of]     
end

module Score = struct
  type score = 
    | HighCard
    | OnePair
    | TwoPair
    | ThreeOfKind
    | FullHouse
    | FourOfKind
    | FiveOfKind
  [@@deriving show, ord]
end

module CardFreq = struct
  open Score
  include Card
  include Comparable.Make_plain(Card)

  (* create a map of frequencies *)
  let rec frequency (cards: card list) map = 
    match cards with
    | [] -> map
    | hd :: tl -> (
      let update = fun maybe_count -> 
        match maybe_count with
        | Some c -> Some (c + 1)
        | None -> Some 1
      in
      let map = Map.change map hd ~f:update in
      frequency tl map
    )

  (* each scoring corresponds to a possible mapping *)
  let score_of_cards (cards: card list) : score =
    let empty = Map.of_alist_exn [] in
    let frequencies = frequency cards empty |> Map.data |> List.sort ~compare:Int.compare in
    match frequencies with
    | 1 :: 1 :: 1 :: 1 :: 1 :: [] -> HighCard
    | 1 :: 1 :: 1 ::      2 :: [] -> OnePair
    | 1 :: 2 :: 2 :: [] -> TwoPair
    | 1 :: 1 :: 3 :: [] -> ThreeOfKind
    | 2 :: 3 :: [] -> FullHouse
    | 1 :: 4 :: [] -> FourOfKind
    | 5 :: [] -> FiveOfKind
    | _ -> assert false
end


module OCamlCard = struct
  open Card
  open Score

  type hand = {
    cards: card list;
    bid: int
  }

  let parse_hand line : hand = 
    let parse_card c : card = 
      match Char.get_digit c with
      | Some v -> Num v
      | None -> (
        match c with
        | 'A' -> A
        | 'K' -> K
        | 'Q' -> Q
        | 'J' -> J
        | 'T' -> T
        | _ -> assert false
      )
    in
    match String.to_list line with
    | a::b::c::d::e::' '::bid -> (
      let cards = List.map ~f:parse_card [a; b; c; d; e] in
      let bid = bid |> of_char_list |> int_of_string in
      {cards; bid}
    )
    | _ -> assert false

  let compare (h1: hand) (h2: hand) : int = 
    let s1 = CardFreq.score_of_cards h1.cards in
    let s2 = CardFreq.score_of_cards h2.cards in
    let score_cmp: int = Score.compare_score s1 s2 in
    if score_cmp = 0 then tiebreaker h1.cards h2.cards else score_cmp
end

open OCamlCard
open Card

let p1_calc hands = 
  hands |> 
    List.sort ~compare:(OCamlCard.compare) |>
    List.mapi ~f:(fun i -> fun h -> (i + 1) * h.bid) |> 
    List.fold_left ~f:(+) ~init:0

let () = 
  let input = In_channel.read_lines "../input.txt" in
  let hands = input |> List.map ~f:parse_hand in
  let p1_ans = p1_calc hands in
    printf "Part 1 answer: %d\n" p1_ans

File: 2023/07/haskell/07.hs
import Data.List

data Card
  = Joker
  | N Int
  | T
  | J
  | Q
  | K
  | A
  deriving (Show, Eq, Ord)

-- this should really be Read
cardParse :: Char -> Card
cardParse 'A' = A
cardParse 'K' = K
cardParse 'Q' = Q
cardParse 'J' = J
cardParse 'T' = T
cardParse c = N (read [c])

data Score
  = HighCard
  | OnePair
  | TwoPair
  | ThreeKind
  | FullHouse
  | FourKind
  | FiveKind
  deriving (Show, Eq, Ord)

freq :: (Ord a) => [a] -> [Int]
freq xs = sort $ map length $ (group . sort) xs

-- assumes proper length of 5 cards
scoreHand :: [Card] -> Score
scoreHand cards =
  case freq cards of
    [1, 1, 1, 1, 1] -> HighCard
    [1, 1, 1, 2] -> OnePair
    [1, 2, 2] -> TwoPair
    [1, 1, 3] -> ThreeKind
    [2, 3] -> FullHouse
    [1, 4] -> FourKind
    [5] -> FiveKind

tiebreakLe :: [Card] -> [Card] -> Bool
tiebreakLe [] [] = True
tiebreakLe (x : xs) (y : ys)
  | x == y = tiebreakLe xs ys
  | otherwise = x <= y

newtype Hand = H [Card] deriving (Show, Eq)

instance Ord Hand where
  (H c1) <= (H c2)
    | s1 == s2 = tiebreakLe c1 c2
    | otherwise = s1 <= s2
    where
      s1 = scoreHand c1
      s2 = scoreHand c2

data Player = P {hand :: Hand, bid :: Int}
  deriving (Show)

parse :: String -> [Player]
parse raw = map parse_player raw_split
  where
    raw_split = map words $ lines raw
    parse_player xs =
      case xs of
        [raw_cards, raw_bid] -> P {hand = H (map cardParse raw_cards), bid = read raw_bid}

p1 :: [Player] -> Int
p1 xs = sum $ zipWith (*) [1 ..] (map bid $ sortOn hand xs)

main :: IO ()
main = do
  -- raw <- readFile "../test.txt"
  raw <- readFile "../input.txt"
  let input = parse raw
  print (p1 input)
